creusot/tests/should_succeed/ghost/pcell.rs

```diff
@@ -0,0 +1,158 @@
+extern crate creusot_contracts;
+use creusot_contracts::*;
+
+pub struct Id(()); // TODO: use a why3 abstract type instead ?
+
+impl Id {
+    /// Generate a fresh id.
+    ///
+    /// To ensure that it is different from previously generated ids, see [`Self::differentiate`].
+    #[pure]
+    #[requires(true)]
+    #[ensures(true)]
+    fn fresh() -> Self {
+        Self(())
+    }
+
+    /// When creating two `PCell`s, we don't know that their ids are different.
+    ///
+    /// This function allows us to derive exactly that.
+    #[trusted]
+    #[pure]
+    #[ensures(*self == ^self)]
+    #[ensures(^self != *other)]
+    fn differentiate(&mut self, other: &Self) {}
+}
+
+pub struct PCell<T> {
+    id: Id,
+    inner: std::cell::Cell<T>,
+}
+
+pub struct Token<T> {
+    id: Id,
+    value: T,
+}
+impl<T> ShallowModel for Token<T> {
+    type ShallowModelTy = Self;
+    #[logic]
+    #[open]
+    fn shallow_model(self) -> Self {
+        self
+    }
+}
+impl<T> Token<T> {
+    #[trusted]
+    #[ensures(self.id != other.id)]
+    pub fn disjoint(&mut self, other: &Self) {}
+}
+
+impl<T> PCell<T> {
+    #[trusted]
+    #[pure]
+    fn conjure_token() -> Token<T> {
+        loop {}
+    }
+
+    #[trusted]
+    #[ensures(result.0.id == result.1@.id)]
+    #[ensures(result.1@.value == value)]
+    pub fn new(value: T) -> (Self, GhostBox<Token<T>>) {
+        let id = Id::fresh();
+        let token = ghost! {
+            Self::conjure_token()
+        };
+        let this = Self { id, inner: std::cell::Cell::new(value) };
+        (this, token)
+    }
+
+    #[trusted]
+    #[requires(self.id == token.inner().id)]
+    #[ensures(result == token.inner().value)]
+    pub fn get(&self, token: GhostBox<&Token<T>>) -> T
+    where
+        T: Copy,
+    {
+        self.inner.get()
+    }
+
+    #[trusted]
+    #[requires(self.id == (*token.inner()).id)]
+    #[ensures(self.id == (^token.inner()).id)]
+    #[ensures(value == (^token.inner()).value)]
+    pub fn set(&self, token: GhostBox<&mut Token<T>>, value: T)
+    where
+        T: Copy,
+    {
+        self.inner.set(value);
+    }
+
+    #[trusted]
+    #[requires(self.id == (*token.inner()).id)]
+    #[ensures(self.id == (^token.inner()).id)]
+    #[ensures(result == (*token.inner()).value)]
+    #[ensures(value == (^token.inner()).value)]
+    pub fn replace(&self, token: GhostBox<&mut Token<T>>, value: T) -> T {
+        self.inner.replace(value)
+    }
+
+    #[trusted]
+    #[requires(self.id == (*self_t.inner()).id)]
+    #[ensures(self.id == (^self_t.inner()).id)]
```

Comment:
Would doing a logical borrow help here? I don't know if we allow this specific syntax / case but `&mut self.id == &mut self_t.inner()`?

Reply:
Nope, this crashes Creusot with an obscure error... I think we can't create a mutable borrow "out of thin air" in pearlite, we need to reborrow something.